tree/dataframe/src/RDatasetSpec.cxx

```diff
@@ -28,128 +28,196 @@ RDatasetSpec::REntryRange::REntryRange(Long64_t begin, Long64_t end) : fBegin(be
+   return *this;
 }
 
-////////////////////////////////////////////////////////////////////////////
-/// \brief Add a friend tree or chain with the same tree name to the dataset specification.
-/// \param[in] treeName Name of the tree
-/// \param[in] fileNameGlob Single file name or glob expression for the files where the tree(s) are stored
-/// \param[in] alias String to refer to the particular friend
-///
-/// The filename glob supports the same type of expressions as TChain::Add().
-void RDatasetSpec::AddFriend(const std::string &treeName, const std::string &fileNameGlob, const std::string &alias)
+RSpecBuilder &
+RSpecBuilder::WithFriends(const std::string &treeName, const std::string &fileNameGlob, const std::string &alias)
 {
    fFriendInfo.AddFriend(treeName, fileNameGlob, alias);
+   return *this;
 }
 
-////////////////////////////////////////////////////////////////////////////
-/// \brief Add a friend tree or chain with the same tree name to the dataset specification.
-/// \param[in] treeName Name of the tree
-/// \param[in] fileNameGlobs A vector of file names or glob expressions for the files where the trees are stored
-/// \param[in] alias String to refer to the particular friend
-///
-/// The filename glob supports the same type of expressions as TChain::Add().
-void RDatasetSpec::AddFriend(const std::string &treeName, const std::vector<std::string> &fileNameGlobs,
-                             const std::string &alias)
+RSpecBuilder &RSpecBuilder::WithFriends(const std::string &treeName, const std::vector<std::string> &fileNameGlobs,
+                                        const std::string &alias)
 {
    fFriendInfo.AddFriend(treeName, fileNameGlobs, alias);
+   return *this;
 }
 
-////////////////////////////////////////////////////////////////////////////
-/// \brief Add a friend tree or chain (possibly with different tree names) to the dataset specification.
-/// \param[in] treeAndFileNameGlobs A vector of pairs of tree names and their corresponding file names/globs
-/// \param[in] alias String to refer to the particular friend
-///
-/// The filename glob supports the same type of expressions as TChain::Add().
-///
-/// ### Example usage:
-/// ~~~{.py}
-/// spec = ROOT.RDF.Experimental.RDatasetSpec("tree", "file.root")
-/// spec.AddFriend([("tree1", "a.root"), ("tree2", "b.root")], "alias")
-/// df = ROOT.RDataFrame(spec)
-/// ~~~
-void RDatasetSpec::AddFriend(const std::vector<std::pair<std::string, std::string>> &treeAndFileNameGlobs,
-                             const std::string &alias)
+RSpecBuilder &RSpecBuilder::WithFriends(const std::vector<std::pair<std::string, std::string>> &treeAndFileNameGlobs,
+                                        const std::string &alias)
 {
    fFriendInfo.AddFriend(treeAndFileNameGlobs, alias);
+   return *this;
 }
 
-const std::vector<std::string> &RDatasetSpec::GetTreeNames() const
+RSpecBuilder &RSpecBuilder::WithRange(const RDatasetSpec::REntryRange &entryRange)
```

Comment:
Maybe we get rid of this?! And we keep the REntryRange only internal, so that it can help in distrdf